File: backend/models/user.model.go
package models

import (
	"time"

	"github.com/go-playground/validator"
	"github.com/google/uuid"
)

type User struct {
	ID        *uuid.UUID `gorm:"type:uuid;default:uuid_generate_v4();primaryKey"`
	UserName  string     `gorm:"type:varchar(100);uniqueIndex;not null"`
	FirstName string     `gorm:"type:varchar(100);not null"`
	LastName  string     `gorm:"type:varchar(100);not null"`
	Email     string     `gorm:"type:varchar(100);uniqueIndex;not null"`
	Phone     string     `gorm:"type:varchar(20);uniqueIndex;not null"`
	Password  string     `gorm:"type:varchar(100);not null"`
	Verified  *bool      `gorm:"not null;default:false"`
	Role      *string    `gorm:"type:varchar(50);default:'user';not null"`
	CreatedAt *time.Time
	UpdatedAt *time.Time
	// One-to-one relationship with Account
	Account *Account `gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
}

type SignUpInput struct {
	UserName        string `json:"user_name" validate:"required"`
	FirstName       string `json:"first_name" validate:"required"`
	LastName        string `json:"last_name" validate:"required"`
	Email           string `json:"email" validate:"required"`
	Phone           string `json:"phone" validate:"required"`
	Password        string `json:"password" validate:"required,min=8"`
	PasswordConfirm string `json:"password_confirm" validate:"required,min=8"`
}

type SignInInput struct {
	Email    string `json:"email"  validate:"required"`
	Password string `json:"password"  validate:"required"`
}

type SignInByPhone struct {
	Phone    string `json:"phone"  validate:"required"`
	Password string `json:"password"  validate:"required"`
}

type UserResponse struct {
	AccountId uuid.UUID `json:"account_id,omitempty"`
	ID        uuid.UUID `json:"id,omitempty"`
	UserName  string    `json:"user_name,omitempty"`
	FirstName string    `json:"first_name,omitempty"`
	LastName  string    `json:"last_name,omitempty"`
	Email     string    `json:"email,omitempty"`
	Phone     string    `json:"phone,omitempty"`
	Role      string    `json:"role,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type UserUpdate struct {
	UserName  string `json:"user_name,omitempty"`
	FirstName string `json:"first_name,omitempty"`
	LastName  string `json:"last_name,omitempty"`
	Email     string `json:"email,omitempty"`
	Phone     string `json:"phone,omitempty"`
	Password  string `json:"password,omitempty"`
	Role      string `json:"role,omitempty"`
	Verified  bool   `json:"verified,omitempty"`
}

type UserPasswordUpdate struct {
	OldPassword        string `json:"old_password,omitempty"`
	NewPassword        string `json:"new_password,omitempty"`
	ConfirmNewPassword string `json:"confirm_new_password,omitempty"`
}

type SignInResponse struct {
	Token       string       `json:"token"`
	UserDetails UserResponse `json:"user"`
}

func FilterUserRecord(user *User, accountId *uuid.UUID) UserResponse {
	var accountID uuid.UUID

	// Check if accountId is nil, handle accordingly
	if accountId != nil {
		accountID = *accountId
	}

	return UserResponse{
		AccountId: accountID, // AccountId is nil if no account
		ID:        *user.ID,
		UserName:  user.UserName,
		FirstName: user.FirstName,
		LastName:  user.LastName,
		Email:     user.Email,
		Phone:     user.Phone,
		Role:      *user.Role,
		CreatedAt: *user.CreatedAt, // Ensure non-nil value
		UpdatedAt: *user.UpdatedAt, // Ensure non-nil value
	}
}

var validate = validator.New()

type ErrorResponse struct {
	Field string `json:"field"`
	Tag   string `json:"tag"`
	Value string `json:"value,omitempty"`
}

func ValidateStruct[T any](payload T) []*ErrorResponse {
	var errors []*ErrorResponse
	err := validate.Struct(payload)
	if err != nil {
		for _, err := range err.(validator.ValidationErrors) {
			var element ErrorResponse
			element.Field = err.StructNamespace()
			element.Tag = err.Tag()
			element.Value = err.Param()
			errors = append(errors, &element)
		}
	}
	return errors
}

File: wallet_app/lib/models/user/user_model.dart
class UserDetails {
  final String accountId;
  final String id;
  final String userName;
  final String firstName;
  final String lastName;
  final String email;
  final String phone;
  final String role;

  UserDetails({
    required this.accountId,
    required this.id,
    required this.userName,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phone,
    required this.role,
  });

  // Convert a UserDetails from a Map
  factory UserDetails.fromJson(Map<String, dynamic> json) {
    return UserDetails(
      accountId: json['account_id'] as String,
      id: json['id'] as String,
      userName: json['user_name'] as String,
      firstName: json['first_name'] as String,
      lastName: json['last_name'] as String,
      email: json['email'] as String,
      phone: json['phone'] as String,
      role: json['role'] as String,
    );
  }

  // Convert a UserDetails to a Map
  Map<String, dynamic> toJson() {
    return {
      'account_id': accountId,
      'id': id,
      'user_name': userName,
      'first_name': firstName,
      'last_name': lastName,
      'email': email,
      'phone': phone,
      'role': role,
    };
  }

  factory UserDetails.fromUserResponse(UserResponse userResponse) {
    return UserDetails(
      accountId: userResponse.accountId,
      id: userResponse.id,
      userName: userResponse.userName,
      firstName: userResponse.firstName,
      lastName: userResponse.lastName,
      email: userResponse.email,
      phone: userResponse.phone,
      role: userResponse.role,
    );
  }
}

class UserResponse {
  final String accountId;
  final String id;
  final String userName;
  final String firstName;
  final String lastName;
  final String email;
  final String phone;
  final String role;
  final DateTime createdAt;
  final DateTime updatedAt;

  UserResponse({
    required this.accountId,
    required this.id,
    required this.userName,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phone,
    required this.role,
    required this.createdAt,
    required this.updatedAt,
  });

  // Convert a UserResponse from a Map
  factory UserResponse.fromJson(Map<String, dynamic> json) {
    return UserResponse(
      accountId: json['account_id'] as String,
      id: json['id'] as String,
      userName: json['user_name'] as String,
      firstName: json['first_name'] as String,
      lastName: json['last_name'] as String,
      email: json['email'] as String,
      phone: json['phone'] as String,
      role: json['role'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  // Convert a UserResponse to a Map
  Map<String, dynamic> toJson() {
    return {
      'account_id': accountId,
      'id': id,
      'user_name': userName,
      'first_name': firstName,
      'last_name': lastName,
      'email': email,
      'phone': phone,
      'role': role,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  // Convert UserResponse to UserDetails
  UserDetails toUserDetails() {
    return UserDetails(
      accountId: accountId,
      id: id,
      userName: userName,
      firstName: firstName,
      lastName: lastName,
      email: email,
      phone: phone,
      role: role,
    );
  }
}
